traits/src/data_provider.rs

```diff
@@ -0,0 +1,138 @@
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+/// Data provider with ability to provide data with no-op, and provide all data.
+pub trait DataFeeder<Key, Value, AccountId>: DataProvider<Key, Value> {
+	/// Provide a new value for a given key from an operator
+	fn feed_value(who: AccountId, key: Key, value: Value) -> DispatchResult;
+}
+
+/// A simple trait to provide data
+pub trait DataProvider<Key, Value> {
+	/// Get data by key
+	fn get(key: &Key) -> Option<Value>;
+}
+
+/// Extended data provider to provide timestamped data by key with no-op, and
+/// all data.
+pub trait DataProviderExtended<Key, TimestampedValue> {
+	/// Get timestamped value by key
+	fn get_no_op(key: &Key) -> Option<TimestampedValue>;
+	/// Provide a list of tuples of key and timestamped value
+	fn get_all_values() -> Vec<(Key, Option<TimestampedValue>)>;
+}
+
+#[allow(dead_code)] // rust cannot defect usage in macro_rules
+pub fn median<T: Ord + Clone>(mut items: Vec<T>) -> Option<T> {
+	if items.is_empty() {
+		return None;
+	}
+
+	let mid_index = items.len() / 2;
+	// Won't panic as guarded items not empty case.
+	let (_, median, _) = items.partition_at_index(mid_index);
+	Some(median.clone())
+}
+
+#[macro_export]
+macro_rules! create_median_value_data_provider {
+	($name:ident, $key:ty, $value:ty, $timestamped_value:ty, [$( $provider:ty ),*]) => {
+		pub struct $name;
+		impl $crate::DataProvider<$key, $value> for $name {
+			fn get(key: &$key) -> Option<$value> {
+				let mut values = vec![];
```

Comment:
```suggestion
				let values = vec![ $(<$provider as $crate::DataProvider<$key, $value>>::get(&key)),* ];
```

Reply:
The `if let` below are to filter out `None` values, or we have to do a `retain`.